physical/consul.go

```diff
@@ -84,14 +160,247 @@ func newConsulBackend(conf map[string]string) (Backend, error) {
+	if !c.disableRegistration && atomic.CompareAndSwapInt64(&c.registrationLock, 0, 1) {
+		defer atomic.CompareAndSwapInt64(&c.registrationLock, 1, 0)
+
+		// Retry agent registration until successful
+		for {
+			c.service.Tags = serviceTags(c.active)
+			agent := c.client.Agent()
+			err := agent.ServiceRegister(c.service)
+			if err == nil {
+				// Success
+				return nil
+			}
+
+			// wtb logger c.logger.Printf("[WARN] service registration failed: %v", err)
+			c.serviceLock.Unlock()
+			time.Sleep(registrationRetryInterval)
+			c.serviceLock.Lock()
+
+			if !c.running {
+				// Shutting down
+				return err
+			}
+		}
+	}
+
+	// Successful concurrent update to active state
+	return nil
+}
+
+func (c *ConsulBackend) AdvertiseSealed(sealed bool) error {
+	c.serviceLock.Lock()
+	defer c.serviceLock.Unlock()
+	c.unsealed = !sealed
+
+	// Vault is still bootstrapping
+	if c.service == nil {
+		return nil
+	}
+
+	if !c.disableRegistration {
+		// Push a TTL check immediately to update the state
+		c.runCheck()
+	}
+
+	return nil
+}
+
+func (c *ConsulBackend) setAdvertiseAddr(addr string) (err error) {
+	c.advertiseHost, c.advertisePort, err = c.parseAdvertiseAddr(addr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *ConsulBackend) RunServiceDiscovery(shutdownCh ShutdownChannel, advertiseAddr string) (err error) {
+	c.serviceLock.Lock()
+	defer c.serviceLock.Unlock()
+
+	if c.disableRegistration {
+		return nil
+	}
+
+	if err := c.setAdvertiseAddr(advertiseAddr); err != nil {
+		return err
+	}
+
+	serviceID := c.serviceID()
+
+	c.service = &api.AgentServiceRegistration{
+		ID:                serviceID,
+		Name:              c.serviceName,
+		Tags:              serviceTags(c.active),
+		Port:              c.advertisePort,
+		Address:           c.advertiseHost,
+		EnableTagOverride: false,
+	}
+
+	checkStatus := api.HealthCritical
+	if c.unsealed {
+		checkStatus = api.HealthPassing
+	}
+
+	c.sealedCheck = &api.AgentCheckRegistration{
+		ID:        c.checkID(),
+		Name:      "Vault Sealed Status",
+		Notes:     "Vault service is healthy when Vault is in an unsealed status and can become an active Vault server",
+		ServiceID: serviceID,
+		AgentServiceCheck: api.AgentServiceCheck{
+			TTL:    c.checkTimeout.String(),
+			Status: checkStatus,
+		},
+	}
+
+	agent := c.client.Agent()
+	if err := agent.ServiceRegister(c.service); err != nil {
+		return errwrap.Wrapf("service registration failed: {{err}}", err)
+	}
+
+	if err := agent.CheckRegister(c.sealedCheck); err != nil {
+		return errwrap.Wrapf("service registration check registration failed: {{err}}", err)
+	}
+
+	go c.checkRunner(shutdownCh)
+	c.running = true
+
+	// Deregister upon shutdown
+	go func() {
+	shutdown:
+		for {
+			select {
+			case <-shutdownCh:
+				// wtb logger: log.Printf("[DEBUG]: Shutting down consul backend")
+				break shutdown
+			}
+		}
+
+		if err := agent.ServiceDeregister(serviceID); err != nil {
```

Comment:
Nope.  When the service is deregistered, the associated checks that are associated with the service are automatically GC'ed for us by Consul.